src/shared/ui/picker.ts

```diff
@@ -145,3 +148,213 @@ export function verifySinglePickerOutput<T extends vscode.QuickPickItem>(choices
+            this.quickPick.busy = true
+            try {
+                const newItems = await this.iterator.next()
+                if (!newItems) {
+                    this.isDone = true
+                    break
+                }
+                // TODO: Is there a way to append to this ReadOnlyArray so it doesn't constantly pop focus back to the top?
+                // on reset, first iteration should clear the quickPick's items.
+                // should handle cases where call latencies are stable
+                if (scopeIterationTime === this.activeIterationTime) {
+                    this.quickPick.items = reset ? newItems.value : this.quickPick.items.concat(newItems.value)
+                    // nothing else to reset from here on out
+                    reset = false
+                    this.isDone = newItems.done ?? false
+                } else {
+                    // another newer iteration cycle is in-flight. Break without mutating state.
+                    break
+                }
+            } catch (e) {
+                // append error node
+                // clicking error node should either go backwards (return undefined) or refresh
+                // maybe have one of each?
+                // give quickpick an error message
+                // we should not blow away the existing items, they should still be viable
+                const err = e as Error
+                this.quickPick.items = [
+                    ...this.quickPick.items,
+                    {
+                        ...IteratingQuickPickController.ERROR_ITEM,
+                        detail: err.message,
+                    },
+                ]
+                this.isDone = true
+            }
+        }
+
+        // wrap up, but only if the fucntion call is the currently active one.
+        if (scopeIterationTime === this.activeIterationTime) {
+            // no items in response
+            if (this.isDone && this.quickPick.items.length === 0) {
+                this.quickPick.items = [IteratingQuickPickController.NO_ITEMS_ITEM]
+            }
+
+            // disable loading bar
+            this.quickPick.busy = false
+        }
+    }
+}
+
+/**
+ * Represents an iterator that tranforms another iterator into an array of QuickPickItems.
+ * Additionally, has a reset functionality to reset the iterator to its initial state.
+ */
+export class IteratingQuickPickPopulator<TResponse> {
+    private iterator: AsyncIterator<TResponse>
+
+    /**
+     * @param iteratorFactory Function that returns an iterator, with all default state values set. E.g. `collectionUtils.getPaginatedAwsCallIter`
+     * @param transform Function which transforms a response from the iterator into an array of `vscode.QuickPickItem`s.
+     */
+    public constructor(
+        private readonly iteratorFactory: () => AsyncIterator<TResponse>,
+        private readonly transform: (response: TResponse) => vscode.QuickPickItem[]
```

Comment:
Yes. The iterator factory is an adapter that turns essentially direct AWS calls into paginated iterators. That functionality is currently exclusive to this function but my intention is that that will become wider used by other paginated loaders.

Because of this, this class exists solely to be an adapter from responses to QuickPickItems.

Reply:
I see what you mean.  It sounds like you want a wrapper that takes one iterator and transforms its results into a different type. This is [forwarding](https://en.wikipedia.org/wiki/Forwarding_(object-oriented_programming)) the call. It sounds like this sits at a halfway point between generic and specific to to quickpick. 

Consider that you could make (or find a library for) a `forwardingIterator()` function that applies *any* transformation on the results of an iterator that returns *any* type. This is effectively done by replacing usages of `vscode.QuickPickItem[]` with another *different* generic parameter.